### PR TITLE
Roles fullstack

### LINKED ISSUE #169: Role specific pages (fullstack)
**Branch:** `roles-fullstack`

### Task at Hand

1. Implement your code in `AuthContext.tsx` and any other files necessary.
2. Confirm that your code and pull request follow the [Pull Request Checklist](https://www.notion.so/h4i/Pull-Request-Checklist-910d5e6925df49569067f00f6aa0c0f3).

### Acceptance Criteria

- [x] Volunteers should be able to only access the pages they have access to.
- [x] Supervisors should be able to only access the pages they have access to.
- [x] Admins should be able to only access the pages they have access to.
- [x] When a user’s role is changed, update their Firebase custom claim (see `ManageUserProfile.tsx`)

### Notes

- Use Firebase custom tokens to check the role of the user and see whether they are volunteers, supervisors, or admins. Only allow users with the necessary permission to view specific pages.
- Supervisors only: Manage Event, Create/Edit Event, View Events (only part of the website)
- Admins: Manage Website (not implemented yet), Manage Members, FAQ page (only part of the webiste)

### PR DESCRIPTION
**Pull request by owenchen04**
## Summary

Used firebase custom claims to restrict which roles have access to certain pages. Added role field in AuthContext dependent on the custom claim values in order to conditionally display content in ViewEvents.tsx and About.tsx. Updated users/views.ts to update custom claims in firebase when a user's role is changed.

## Testing

We did manual testing by creating new users with different roles and checking which pages they had access to. We also changed their roles on the ManageUserProfileNew.tsx page, then checked to see that their page access had changed correspondingly. For the conditional rendering on ViewEvents and About, we added some divs with dummy content that were conditionally rendered based on the role, and logged into an account for each different role to see that they were shown the correct content.

## Notes

For ViewEvents and About, the actual content to be conditionally rendered will be implemented in later tickets, so for now they just have dummy divs to show that the custom claims work.

Closes #169

### CODE REVIEW

File: frontend/src/utils/AuthContext.tsx [hunk 1]
```diff
@@ -113,17 +117,70 @@ export const AuthProvider = ({ children }: AuthProviderProps) => {
...
+            if (!adminPaths.includes(path) && !supervisorPaths.includes(path)) {
+              setIsAuthenticated(true);
+            } else {
+              router.replace("/events/view");
+            }
+          } else {
+            router.replace("/login");
+            setIsAuthenticated(false);
+          }
+        } else {
```
Comment (akinfelami) [inline frontend/src/utils/AuthContext.tsx:171]: Not a huge fan of chained if statements, especially in a useEffect. See if you can tidy this logic up a little bit. It is almost non-ideal for authentication where things have to be deterministic. I recommended simplifying some of the conditions using `&&` or using case statements!

**Review by akinfelami — CHANGES_REQUESTED** [review]
Hey guys, 

Amazing work on this so far!!

Like I said, there are a few things to tidy up. Start by making all the functions in `auth.ts` to handle promises properly, then you can test that it actually works. Then you can move to addressing changes in the frontend.

File: frontend/src/utils/AuthContext.tsx (hunk 1 above)
Comment (owenchen04) [reply frontend/src/utils/AuthContext.tsx:171]: I combined some of the conditions using <&&> to hopefully make it a bit more readable!

**Comment by akinfelami** [comment]
Folks, my last change addressed code readability and functionality. Kindly take a look at the changes to make sure they all make sense and  Please test again to make sure everything looks good!

**Comment by akinfelami** [comment]
Fix merge conflicts!

**Review by akinfelami — APPROVED** [review]

**Pull request merged**